cmd/compass/main.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/spf13/cobra"
+)
+
+func main() {
+	// Parent command to which all subcommands are added.
+	cmds := &cobra.Command{
+		Use:   "compass",
+		Short: "Interact with OCI image layouts.",
+		Run: func(cmd *cobra.Command, args []string) {
+			cmd.Help()
+		},
+	}
+
+	// TODO: rm
+	//  remove an entry from index.json, optionally garbage collecting unreferenced
+	//  blobs
+	// TODO: gc
+	//  garbage collect unreferenced blobs
+	for _, cmd := range []cobra.Command{{
+		Use:   "freeze",
+		Short: "Pull a remote repo and store it in an oci image layout",
+		Args:  cobra.ExactArgs(2),
+		Run: func(_ *cobra.Command, args []string) {
+			freeze(args[0], args[1])
+		},
+	}, {
+		Use:   "thaw",
+		Short: "Push the images in an image layout to a remote registry",
+		Args:  cobra.ExactArgs(2),
+		Run: func(_ *cobra.Command, args []string) {
+			thaw(args[0], args[1])
+		},
+	}, {
+		Use:   "ls",
+		Short: "List the manifests in a layout",
+		Args:  cobra.ExactArgs(1),
+		Run: func(_ *cobra.Command, args []string) {
+			ls(args[0])
+		},
+	}} {
+		cmd := cmd
+		cmds.AddCommand(&cmd)
+	}
+
+	if err := cmds.Execute(); err != nil {
+		log.Fatalf("error during command execution: %v", err)
+	}
+}
+
+func freeze(src, path string) {
+	repo, err := name.NewRepository(src, name.WeakValidation)
+	if err != nil {
+		log.Fatalf("parsing repo %q: %v", src, err)
+	}
+	log.Printf("Pulling %v", repo)
+
+	auth, err := authn.DefaultKeychain.Resolve(repo.Registry)
+	if err != nil {
+		log.Fatalf("getting creds for %q: %v", repo, err)
+	}
+
+	tags, err := remote.List(repo, auth, http.DefaultTransport)
+	if err != nil {
+		log.Fatalf("reading repo %q: %v", repo, err)
+	}
+
+	for _, tag := range tags {
+		t := fmt.Sprintf("%s:%s", repo, tag)
+		ref, err := name.NewTag(t, name.StrictValidation)
+		if err != nil {
+			log.Fatalf("parsing tag %q: %v", t, err)
+		}
+		log.Printf("Pulling %v", ref)
+		img, err := remote.Image(ref, remote.WithAuth(auth))
+		if err != nil {
+			log.Fatalf("reading image %q: %v", ref, err)
+		}
+
+		annotations := map[string]string{
+			"org.opencontainers.image.ref.name": t,
+		}
+		if _, err := layout.AppendImage(path, img, layout.WithAnnotations(annotations)); err != nil {
+			log.Fatalf("writing image %q: %v", path, err)
+		}
+	}
+}
+
+func thaw(path, dst string) {
+	repo, err := name.NewRepository(dst, name.WeakValidation)
+	if err != nil {
+		log.Fatalf("parsing repo %q: %v", dst, err)
+	}
+	log.Printf("Pushing %v", repo)
+
+	auth, err := authn.DefaultKeychain.Resolve(repo.Registry)
+	if err != nil {
+		log.Fatalf("getting creds for %q: %v", repo, err)
+	}
+
+	ii, err := layout.Index(path)
+	if err != nil {
+		log.Fatalf("reading image layout %q: %v", path, err)
+	}
+
+	manifest, err := ii.IndexManifest()
+	if err != nil {
+		log.Fatalf("reading image layout manifest %q: %v", path, err)
+	}
+
+	for _, desc := range manifest.Manifests {
+		// This logic is janky, we'd want smarter rewriting rules.
```

Comment:
Can you describe some of the smarts in case someone else is responsible for adding them?